File: packages/project-ledger/src/records.js
import { existsSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { basename, dirname, join, relative } from "node:path";
import {
  PRIVATE_CONTENT_PATTERNS,
  VALID_ATTEMPT_STATES,
  VALID_TASK_STATES,
  VALID_WORK_STATES,
} from "./constants.js";
import { CliError, nowIso } from "./errors.js";
import { ensureDir, ledgerRoot, listFiles, projectRelative, safeReadJson } from "./fs.js";
import { frontmatterBody, markdownWithFrontmatter, parseFrontmatter } from "./frontmatter.js";
import { completionGateIssues } from "./state-machine.js";
import { isNonAuthoritativeProjectLedgerStorage } from "./storage-authority.js";

export function issue(code, severity, message, path, record = null) {
  return {
    code,
    severity,
    message,
    path,
    record: record ? recordReference(record) : null,
  };
}

export function recordReference(record, reason = null) {
  const ref = {
    id: record.id,
    kind: record.kind,
    title: record.title,
    status: record.status,
    path: record.path,
  };
  if (reason) ref.reason = reason;
  return ref;
}

export function inferKind(relPath, data) {
  const ledgerPath = ledgerContentPath(relPath);
  if (typeof data.kind === "string" && data.kind) return data.kind;
  if (ledgerPath === "project.json") return "project";
  if (/^work\/[^/]+\/tasks\/[^/]+\/attempts\//u.test(ledgerPath)) return "attempt";
  if (/^work\/[^/]+\/tasks\//u.test(ledgerPath)) return "task";
  if (ledgerPath.startsWith("work/")) return "work";
  if (ledgerPath.startsWith("initiatives/")) return "initiative";
  if (ledgerPath.startsWith("decisions/")) return "decision";
  if (ledgerPath.startsWith("risks/")) return "risk";
  if (ledgerPath.startsWith("specs/")) return "spec";
  if (ledgerPath.startsWith("reports/")) return "report";
  if (ledgerPath.startsWith("plans/")) return "plan";
  if (ledgerPath.startsWith("handoffs/")) return "handoff";
  if (ledgerPath.startsWith("references/")) return "reference";
  if (ledgerPath.startsWith("roadmaps/")) return "roadmap";
  return "record";
}

export function readRecordData(filePath) {
  const text = readFileSync(filePath, "utf8");
  if (filePath.endsWith(".json")) return safeReadJson(filePath);
  if (filePath.endsWith(".md")) return parseFrontmatter(text) ?? {};
  return null;
}

export function readRecord(project, filePath) {
  const relPath = projectRelative(project, filePath);
  if (relPath.endsWith("ledger.jsonl")) return null;
  if (isNonAuthoritativeProjectLedgerStorage(filePath)) return null;
  const stats = statSync(filePath);
  const data = readRecordData(filePath);
  if (!data) return null;

  const semanticPath = relative(ledgerRoot(project), filePath).split("\\").join("/");
  const kind = inferKind(semanticPath, data);
  const id = typeof data.id === "string" && data.id.trim()
    ? data.id.trim()
    : basename(filePath).replace(/\.(json|md)$/u, "");
  return {
    id,
    kind,
    title: String(data.title ?? data.name ?? id),
    status: String(data.status ?? "unknown"),
    priority: typeof data.priority === "number" ? data.priority : 100,
    parentId: typeof data.parentId === "string" ? data.parentId : null,
    spec: typeof data.spec === "string" ? data.spec : null,
    specExemption: Boolean(data.specExemption),
    acceptance: typeof data.acceptance === "string" ? data.acceptance : null,
    acceptanceExemption: Boolean(data.acceptanceExemption),
    requiresCommitEvidence: Boolean(data.requiresCommitEvidence),
    codeCommits: typeof data.codeCommits === "string" ? data.codeCommits : null,
    ledgerCommits: typeof data.ledgerCommits === "string" ? data.ledgerCommits : null,
    report: typeof data.report === "string" ? data.report : null,
    review: typeof data.review === "string" ? data.review : null,
    validation: typeof data.validation === "string" ? data.validation : null,
    implementation: typeof data.implementation === "string" ? data.implementation : null,
    mitigation: typeof data.mitigation === "string" ? data.mitigation : null,
    reason: typeof data.reason === "string" ? data.reason : null,
    updatedAt: String(data.updatedAt ?? new Date(stats.mtimeMs).toISOString()),
    path: relPath,
    sourceMtimeMs: stats.mtimeMs,
  };
}

export function readRecordBody(filePath) {
  if (!filePath.endsWith(".md")) return null;
  return frontmatterBody(readFileSync(filePath, "utf8"));
}

export function recordFiles(project) {
  const root = ledgerRoot(project);
  const projectFile = join(root, "project.json");
  return [
    projectFile,
    ...listFiles(root, { skipDirs: new Set(["index", "views"]) })
      .filter((file) => file !== projectFile && !file.endsWith("ledger.jsonl") &&
        !isNonAuthoritativeProjectLedgerStorage(file)),
  ].filter((file) => existsSync(file));
}

export function sourceMaxMtimeMs(project) {
  return recordFiles(project).reduce((max, file) => Math.max(max, statSync(file).mtimeMs), 0);
}

export function validateRecord(record, recordsById) {
  const issues = [];
  if (!record.id || record.id === "work" || record.id === "project") {
    issues.push(issue("invalid_schema", "error", "Record id is missing or too generic", record.path, record));
  }
  if (record.kind === "work" && !VALID_WORK_STATES.has(record.status)) {
    issues.push(issue("invalid_state", "error", `Invalid work state: ${record.status}`, record.path, record));
  }
  if (record.kind === "task" && !VALID_TASK_STATES.has(record.status)) {
    issues.push(issue("invalid_state", "error", `Invalid task state: ${record.status}`, record.path, record));
  }
  if (record.kind === "attempt" && !VALID_ATTEMPT_STATES.has(record.status)) {
    issues.push(issue("invalid_state", "error", `Invalid attempt state: ${record.status}`, record.path, record));
  }
  if (record.kind === "task" && record.parentId && !recordsById.has(record.parentId)) {
    issues.push(issue("orphan_task", "error", `Task parent does not exist: ${record.parentId}`, record.path, record));
  }
  if (
    record.kind === "work" &&
    !["done", "cancelled"].includes(record.status) &&
    !record.spec &&
    !record.specExemption
  ) {
    issues.push(issue("missing_spec", "warning", "Active work has no linked spec or spec exemption", record.path, record));
  }
  for (const gateIssue of completionGateIssues(record)) {
    issues.push(issue("completion_gate", "error", gateIssue.message, record.path, record));
  }
  return issues;
}

export function scanPrivacy(project) {
  const findings = [];
  for (const file of listFiles(ledgerRoot(project), { skipDirs: new Set(["index", "views"]) })) {
    if (file.endsWith("ledger.jsonl")) continue;
    if (isNonAuthoritativeProjectLedgerStorage(file)) continue;
    const text = readFileSync(file, "utf8");
    if (PRIVATE_CONTENT_PATTERNS.some((pattern) => pattern.test(text))) {
      findings.push(issue(
        "possible_private_content",
        "error",
        "Record may contain private content; inspect locally without copying raw text.",
        projectRelative(project, file),
      ));
    }
  }
  return findings;
}

export function workRecordPath(project, id) {
  return join(ledgerRoot(project), "work", safeRecordId(id), "work.md");
}

export function topLevelRecordPath(project, kind, id) {
  const dir = {
    initiative: "initiatives",
    decision: "decisions",
    risk: "risks",
    spec: "specs",
    report: "reports",
    plan: "plans",
    handoff: "handoffs",
    reference: "references",
    roadmap: "roadmaps",
  }[kind];
  if (!dir) throw new CliError(`Unsupported top-level record kind: ${kind}`, "invalid_input", 1);
  return join(ledgerRoot(project), dir, `${safeRecordId(id).toLowerCase()}.md`);
}

export function taskRecordPath(project, workId, id) {
  return join(ledgerRoot(project), "work", safeRecordId(workId), "tasks", `${safeRecordId(id)}.md`);
}

export function attemptRecordPath(project, workId, taskId, id) {
  return join(
    ledgerRoot(project),
    "work",
    safeRecordId(workId),
    "tasks",
    safeRecordId(taskId),
    "attempts",
    `${safeRecordId(id)}.md`,
  );
}

export function writeMarkdownRecord(filePath, data, body = null) {
  ensureDir(dirname(filePath));
  const text = body ?? `# ${data.title ?? data.id}\n\nManaged by Project Ledger.`;
  writeFileSync(filePath, markdownWithFrontmatter({
    ...data,
    updatedAt: nowIso(),
  }, text), "utf8");
}

export function updateMarkdownRecord(filePath, updates, bodyOverride = undefined) {
  const existingText = readFileSync(filePath, "utf8");
  const existing = readRecordData(filePath) ?? {};
  const body = bodyOverride ?? (frontmatterBody(existingText) || `# ${updates.title ?? existing.title ?? existing.id}\n`);
  writeMarkdownRecord(filePath, { ...existing, ...updates }, body);
}

export function findRecord(index, kind, id) {
  const matches = index.records.filter((record) => record.kind === kind && record.id === id);
  if (matches.length === 0) throw new CliError(`${kind} not found: ${id}`, "record_not_found", 1);
  if (matches.length > 1) throw new CliError(`${kind} id is ambiguous: ${id}`, "ambiguous_record", 1);
  return matches[0];
}

export function workIdFromTaskPath(path) {
  const match = ledgerContentPath(path).match(/^work\/([^/]+)\/tasks\//u);
  return match?.[1] ?? null;
}

export function taskIdFromAttemptPath(path) {
  const match = ledgerContentPath(path).match(/^work\/([^/]+)\/tasks\/([^/]+)\/attempts\//u);
  return match ? { workId: match[1], taskId: match[2] } : null;
}

function ledgerContentPath(path) {
  const normalized = path.split("\\").join("/");
  const canonicalMatch = normalized.match(/^project-ledger\/projects\/[^/]+(?:\/(.*))?$/u);
  if (canonicalMatch) return canonicalMatch[1] ?? "";
  return normalized;
}

function safeRecordId(id) {
  const value = String(id ?? "").trim();
  if (!value || value === "." || value === ".." || value.includes("/") || value.includes("\\")) {
    throw new CliError(`Invalid record id: ${value || "(empty)"}`, "invalid_input", 1);
  }
  return value;
}

File: packages/butler-agent/rust/agent/src/project_ledger/commands/index/records.rs
use std::fs;
use std::path::Path;

use serde_json::{Value, json};

use super::super::{CliFailure, io_failure, record};
use crate::project_ledger::committed;

pub(super) struct Records {
    pub projected: Vec<Value>,
    pub issues: Vec<Value>,
    pub max_mtime_ms: f64,
    pub project: Value,
}

pub(super) fn read(root: &Path) -> Result<Records, CliFailure> {
    let project_path = root.join("project.json");
    if !root.exists() {
        return Err(CliFailure::new(
            "not_initialized",
            "Project Ledger is not initialized",
        ));
    }
    if !project_path.exists() {
        return Err(CliFailure::new(
            "missing_project",
            "project.json is missing",
        ));
    }
    let project_raw = fs::read_to_string(&project_path).map_err(|_| io_failure())?;
    let project_json: Value = serde_json::from_str(&project_raw)
        .map_err(|_| CliFailure::new("invalid_json", "Invalid Project Ledger project JSON"))?;
    let mut projected = Vec::new();
    let mut issues = Vec::new();
    let mut max_mtime_ms = 0.0_f64;
    let mut source_rows = vec![("project.json".to_owned(), Some(project_raw))];
    source_rows.extend(committed::read_all(root).map_err(|_| io_failure())?);
    source_rows.sort_by(|a, b| a.0.cmp(&b.0));
    for (relative, raw) in source_rows {
        let Some(raw) = raw else { continue };
        match record::from_raw(root, Path::new(&relative), &raw) {
            Ok(Some(record)) => {
                max_mtime_ms = max_mtime_ms.max(record.source_mtime_ms);
                projected.push(record.value);
            }
            Ok(None) => {}
            Err(error) => issues.push(issue(
                "invalid_schema",
                "error",
                &format!("Cannot parse record: {}", error.message),
                &super::super::display_path(root, Path::new(&relative)),
                None,
            )),
        }
    }
    let by_id: std::collections::HashMap<_, _> = projected
        .iter()
        .filter_map(|record| {
            record
                .get("id")
                .and_then(Value::as_str)
                .map(|id| (id, record))
        })
        .collect();
    for record in &projected {
        issues.extend(super::validation::validate(record, &by_id));
    }
    let project = projected
        .iter()
        .find(|record| record.get("kind").and_then(Value::as_str) == Some("project"))
        .map(|record| {
            json!({
                "id":record.get("id"),
                "name":record.get("title"),
                "status":record.get("status"),
                "path":record.get("path"),
            })
        })
        .unwrap_or(project_json);
    Ok(Records {
        projected,
        issues,
        max_mtime_ms,
        project,
    })
}

pub(super) fn count(records: &[Value]) -> Value {
    let mut counts = json!({
        "records":records.len(),"initiative":0,"work":0,"task":0,"attempt":0,
        "decision":0,"risk":0,"spec":0,"report":0,"plan":0,
        "handoff":0,"reference":0,"roadmap":0,
    });
    for record in records {
        if let Some(kind) = record.get("kind").and_then(Value::as_str)
            && let Some(value) = counts.get(kind).and_then(Value::as_u64)
        {
            counts[kind] = json!(value + 1);
        }
    }
    counts
}

pub(super) fn issue(
    code: &str,
    severity: &str,
    message: &str,
    path: &str,
    record: Option<&Value>,
) -> Value {
    json!({
        "code":code,"severity":severity,"message":message,"path":path,
        "record":record.map(|record| record::reference(record, None)),
    })
}
